Check dependencies

molant opened the pull request:
<!--

Read our pull request guide:
https://webhint.io/docs/contributor-guide/getting-started/pull-requests/

For the following items put an "x" between the square brackets
(i.e. [x]) if you completed the associated item.

-->

## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- ~~Added/Updated related documentation.~~
- ~~Added/Updated related tests.~~

## Short description of the change(s)

Created `lint-dependencies.js` that reads all the `ts` and `js` files inside each package and checks that the `package.json` has that dependency.

It also adds a task `lint:dependencies` inside each package that calls this script and executes it just in the context of the package. 

There are 2 dependencies that are ignored right now: `estree` and `har-format`. These are `@types` and while we could also check for `@types/packageName` I prefer to do it this way just in case we add the dependency, the type, and then remove the dependency and we are still pulling it from another one.

The current output of the script on my machine:

```
C:\projects\@webhintio\hint>node scripts\lint-dependencies.js
Packages found: 74
No files to process for "C:\projects\@webhintio\hint\packages\configuration-development"
No files to process for "C:\projects\@webhintio\hint\packages\configuration-progressive-web-apps"
No files to process for "C:\projects\@webhintio\hint\packages\configuration-web-recommended"
Processing "C:\projects\@webhintio\hint\packages\connector-edge"
Processing "C:\projects\@webhintio\hint\packages\connector-chrome"
Processing "C:\projects\@webhintio\hint\packages\connector-jsdom"
Processing "C:\projects\@webhintio\hint\packages\connector-local"
Processing "C:\projects\@webhintio\hint\packages\create-hint"
        C:/projects/@webhintio/hint/packages/create-hint/tests/create-hint.ts
                proxyquire
Processing "C:\projects\@webhintio\hint\packages\create-hintrc"
        C:/projects/@webhintio/hint/packages/create-hintrc/tests/create-hintrc.ts
                proxyquire
Processing "C:\projects\@webhintio\hint\packages\create-parser"
        C:/projects/@webhintio/hint/packages/create-parser/tests/new-parser.ts
                proxyquire
Processing "C:\projects\@webhintio\hint\packages\extension-browser"
        C:/projects/@webhintio/hint/packages/extension-browser/tests/background-script.ts
                proxyquire
        C:/projects/@webhintio/hint/packages/extension-browser/tests/content-script.ts
                proxyquire
Processing "C:\projects\@webhintio\hint\packages\extension-vscode"
Processing "C:\projects\@webhintio\hint\packages\formatter-codeframe"
Processing "C:\projects\@webhintio\hint\packages\formatter-excel"
        C:/projects/@webhintio/hint/packages/formatter-excel/tests/tests.ts
                fs-extra
Processing "C:\projects\@webhintio\hint\packages\formatter-html"
Processing "C:\projects\@webhintio\hint\packages\formatter-json"
Processing "C:\projects\@webhintio\hint\packages\formatter-stylish"
Processing "C:\projects\@webhintio\hint\packages\formatter-summary"
        C:/projects/@webhintio/hint/packages/formatter-summary/src/formatter.ts
                strip-ansi
        C:/projects/@webhintio/hint/packages/formatter-summary/tests/tests.ts
                strip-ansi
Processing "C:\projects\@webhintio\hint\packages\hint"
Processing "C:\projects\@webhintio\hint\packages\hint-amp-validator"
Processing "C:\projects\@webhintio\hint\packages\hint-babel-config"
Processing "C:\projects\@webhintio\hint\packages\hint-axe"
Processing "C:\projects\@webhintio\hint\packages\hint-apple-touch-icons"
Processing "C:\projects\@webhintio\hint\packages\hint-compat-api"
Processing "C:\projects\@webhintio\hint\packages\hint-button-type"
Processing "C:\projects\@webhintio\hint\packages\hint-content-type"
Processing "C:\projects\@webhintio\hint\packages\hint-create-element-svg"
Processing "C:\projects\@webhintio\hint\packages\hint-css-prefix-order"
Processing "C:\projects\@webhintio\hint\packages\hint-disown-opener"
Processing "C:\projects\@webhintio\hint\packages\hint-doctype"
Processing "C:\projects\@webhintio\hint\packages\hint-highest-available-document-mode"
Processing "C:\projects\@webhintio\hint\packages\hint-html-checker"
        C:/projects/@webhintio/hint/packages/hint-html-checker/src/hint.ts
                request
Processing "C:\projects\@webhintio\hint\packages\hint-http-cache"
Processing "C:\projects\@webhintio\hint\packages\hint-http-compression"
Processing "C:\projects\@webhintio\hint\packages\hint-https-only"
Processing "C:\projects\@webhintio\hint\packages\hint-manifest-app-name"
Processing "C:\projects\@webhintio\hint\packages\hint-image-optimization-cloudinary"
Processing "C:\projects\@webhintio\hint\packages\hint-manifest-exists"
Processing "C:\projects\@webhintio\hint\packages\hint-manifest-file-extension"
Processing "C:\projects\@webhintio\hint\packages\hint-manifest-is-valid"
Processing "C:\projects\@webhintio\hint\packages\hint-meta-charset-utf-8"
Processing "C:\projects\@webhintio\hint\packages\hint-meta-theme-color"
Processing "C:\projects\@webhintio\hint\packages\hint-meta-viewport"
Processing "C:\projects\@webhintio\hint\packages\hint-minified-js"
Processing "C:\projects\@webhintio\hint\packages\hint-no-bom"
Processing "C:\projects\@webhintio\hint\packages\hint-no-broken-links"
        C:/projects/@webhintio/hint/packages/hint-no-broken-links/src/hint.ts
                request
Processing "C:\projects\@webhintio\hint\packages\hint-no-disallowed-headers"
Processing "C:\projects\@webhintio\hint\packages\hint-no-html-only-headers"
Processing "C:\projects\@webhintio\hint\packages\hint-no-friendly-error-pages"
Processing "C:\projects\@webhintio\hint\packages\hint-no-http-redirects"
Processing "C:\projects\@webhintio\hint\packages\hint-no-protocol-relative-urls"
Processing "C:\projects\@webhintio\hint\packages\hint-no-p3p"
Processing "C:\projects\@webhintio\hint\packages\hint-performance-budget"
Processing "C:\projects\@webhintio\hint\packages\hint-sri"
Processing "C:\projects\@webhintio\hint\packages\hint-no-vulnerable-javascript-libraries"
Processing "C:\projects\@webhintio\hint\packages\hint-ssllabs"
Processing "C:\projects\@webhintio\hint\packages\hint-strict-transport-security"
Processing "C:\projects\@webhintio\hint\packages\hint-stylesheet-limits"
Processing "C:\projects\@webhintio\hint\packages\hint-typescript-config"
Processing "C:\projects\@webhintio\hint\packages\hint-validate-set-cookie-header"
Processing "C:\projects\@webhintio\hint\packages\hint-webpack-config"
Processing "C:\projects\@webhintio\hint\packages\parser-babel-config"
        C:/projects/@webhintio/hint/packages/parser-babel-config/tests/tests.ts
                eventemitter2
Processing "C:\projects\@webhintio\hint\packages\hint-x-content-type-options"
Processing "C:\projects\@webhintio\hint\packages\parser-css"
        C:/projects/@webhintio/hint/packages/parser-css/tests/basic-tests.ts
                eventemitter2
        C:/projects/@webhintio/hint/packages/parser-css/tests/interface-tests.ts
                eventemitter2
Processing "C:\projects\@webhintio\hint\packages\parser-html"
Processing "C:\projects\@webhintio\hint\packages\parser-javascript"
        C:/projects/@webhintio/hint/packages/parser-javascript/tests/tests.ts
                eventemitter2
Processing "C:\projects\@webhintio\hint\packages\parser-manifest"
        C:/projects/@webhintio/hint/packages/parser-manifest/tests/tests.ts
                eventemitter2
Processing "C:\projects\@webhintio\hint\packages\parser-typescript-config"
        C:/projects/@webhintio/hint/packages/parser-typescript-config/tests/tests.ts
                eventemitter2
Processing "C:\projects\@webhintio\hint\packages\parser-package-json"
        C:/projects/@webhintio/hint/packages/parser-package-json/tests/tests.ts
                eventemitter2
Processing "C:\projects\@webhintio\hint\packages\parser-webpack-config"
        C:/projects/@webhintio/hint/packages/parser-webpack-config/src/parser.ts
                webpack
        C:/projects/@webhintio/hint/packages/parser-webpack-config/tests/webpack-config.ts
                eventemitter2
Processing "C:\projects\@webhintio\hint\packages\utils-connector-tools"
Processing "C:\projects\@webhintio\hint\packages\utils-create-server"
Processing "C:\projects\@webhintio\hint\packages\utils-debugging-protocol-common"
        C:/projects/@webhintio/hint/packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts
                lockfile
Processing "C:\projects\@webhintio\hint\packages\utils-tests-helpers"
```

so we have quite a few missing dependencies... It will be interesting in the future to add the inverse check: make sure that all the dependencies are being used, but I'll leave that for another time.

You can start looking into it while I add the missing dependencies.


<!--

If this is a non-trivial change, include information such as what
benefits this change brings as well as possible drawbacks.

If this fixes an existing issue, include the relevant issue number(s).

Thank you for taking the time to open this PR!

-->

sarvaje commented on scripts/lint-dependencies.js:
```diff
@@ -0,0 +1,145 @@
+
+    console.log(`Packages found: ${pkgs.length}`);
+
+    return pkgs.map((pkg) => {
```
if you use the option `absolute: true` in `globby` you won't need this `.map`.

molant replied:
Updated!

molant commented:
I've improved the script to have a list of dependencies should be of `@types` only if the regular package is not as a dependency. THis covers our current packages.

I've added all the missing dependencies found by the script, some missing types I realized during the process (there might be more missing types though) and updated the templates of `create-` packages.

This should be ready to review and I'll be closing #2116 in favor of this one.

antross commented on scripts/lint-dependencies.js:
```diff
@@ -0,0 +1,147 @@
+
+const getCodeContent = async (pkgPath) => {
+    const files = await globby([
+        `${pkgPath}/**/*.ts`,
```
Planning ahead for once we start using React (may also want `jsx`):
```suggestion
        `${pkgPath}/**/*.ts`,
        `${pkgPath}/**/*.tsx`,
```

molant replied:
I'll add `tsx` and `jsx`.

molant commented:
@antross I've done all the changes and changed the history. This should be ready to be rebased.

sarvaje commented:
> @sarvaje any other feedback before I merge this?

Nop, you can merge it
